File: 133_cloneGraph.py
# 133. Clone Graph

# Neetcode: https://www.youtube.com/watch?v=mQeF6bN8hMk

"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        
        oldToNew = {}
        
        def clone(node):
  
            if node in oldToNew:
                return oldToNew[node]
            
            copy = Node(node.val)
            oldToNew[node] = copy
            
            for i in range(len(node.neighbors)):
                copy.neighbors.append(clone(node.neighbors[i]))
                
            return copy
        
        return clone(node) if node else None
        
            
        

File: 133_cloneGraph.cpp
// 133. Clone Graph

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};

*/

class Solution {
public:
    unordered_map<Node*, Node*> oldToNew;
    
    Node* cloneGraph(Node* node) {
        
        if(node!=nullptr)
            return clone(node);
        else
            return node;
    }
    
    Node* clone(Node* node){
            
        if(oldToNew.find(node)!=oldToNew.end())
            return oldToNew[node];

        Node* copy = new Node(node->val);
        oldToNew[node] = copy;

        for(int i=0;i<(node->neighbors).size();i++)
            copy->neighbors.push_back(clone((node->neighbors)[i]));

        return copy;
            
        }
};
